x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -477,97 +475,82 @@ void restoreFiles(Store store) throws IOException {
+                        ListenerTimeouts.wrapWithTimeout(threadPool, threadedListener, ccrSettings.getRecoveryActionTimeout(),
+                            ThreadPool.Names.GENERIC, GetCcrRestoreFileChunkAction.NAME));
                 }
-            }
 
-            logger.trace("[{}] completed CCR restore", shardId);
-        }
+                private void writeFileChunk(StoreFileMetaData md,
+                                            GetCcrRestoreFileChunkAction.GetCcrRestoreFileChunkResponse r) throws Exception {
+                    final int actualChunkSize = r.getChunk().length();
+                    logger.trace("[{}] [{}] got response for file [{}], offset: {}, length: {}",
+                        shardId, snapshotId, md.name(), r.getOffset(), actualChunkSize);
+                    final long nanosPaused = ccrSettings.getRateLimiter().maybePause(actualChunkSize);
+                    throttleListener.accept(nanosPaused);
+                    multiFileWriter.incRef();
+                    try (Releasable ignored = multiFileWriter::decRef) {
+                        final boolean lastChunk = r.getOffset() + actualChunkSize >= md.length();
+                        multiFileWriter.writeFileChunk(md, r.getOffset(), r.getChunk(), lastChunk);
+                    } catch (Exception e) {
+                        handleError(md, e);
+                        throw e;
+                    }
+                }
+
+                @Override
+                protected void handleError(StoreFileMetaData md, Exception e) throws Exception {
+                    final IOException corruptIndexException;
+                    if ((corruptIndexException = ExceptionsHelper.unwrapCorruption(e)) != null) {
+                        try {
+                            store.markStoreCorrupted(corruptIndexException);
+                        } catch (IOException ioe) {
+                            logger.warn("store cannot be marked as corrupted", e);
+                        }
+                        throw corruptIndexException;
+                    }
+                    throw e;
+                }
 
-        private void handleError(Store store, Exception e) throws IOException {
-            final IOException corruptIndexException;
-            if ((corruptIndexException = ExceptionsHelper.unwrapCorruption(e)) != null) {
-                try {
-                    store.markStoreCorrupted(corruptIndexException);
-                } catch (IOException ioe) {
-                    logger.warn("store cannot be marked as corrupted", e);
+                @Override
+                public void close() {
+                    multiFileWriter.close();
                 }
-                throw corruptIndexException;
-            } else {
-                ExceptionsHelper.reThrowIfNotNull(e);
-            }
+            };
+            multiFileTransfer.start();
+            restoreFilesFuture.actionGet(ccrSettings.getRecoveryActionTimeout());
```

Comment:
This would allow recovery from remote to only run for 60 seconds before aborting. I think we want to indefinitely wait here and only have the timeouts on the individual requests.

Reply:
Good catch. Addressed in https://github.com/elastic/elasticsearch/pull/44514/commits/bf3e34d457be81f3501dbd75b5b8fe3f88af0b7d.